polkadot/xcm/src/v4/mod.rs

```diff
@@ -0,0 +1,1459 @@
+		use OldResponse::*;
+		Ok(match old {
+			Null => Self::Null,
+			Assets(assets) => Self::Assets(assets.try_into()?),
+			ExecutionResult(result) =>
+				Self::ExecutionResult(result.map(|(num, old_error)| (num, old_error.into()))),
+			Version(version) => Self::Version(version),
+			PalletsInfo(pallet_info) => {
+				let inner = pallet_info
+					.into_iter()
+					.map(TryInto::try_into)
+					.collect::<result::Result<Vec<_>, _>>()?;
+				Self::PalletsInfo(
+					BoundedVec::<PalletInfo, MaxPalletsInfo>::try_from(inner).map_err(|_| ())?,
+				)
+			},
+			DispatchResult(maybe_error) => Self::DispatchResult(maybe_error),
+		})
+	}
+}
+
+/// Information regarding the composition of a query response.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug, TypeInfo)]
+pub struct QueryResponseInfo {
+	/// The destination to which the query response message should be send.
+	pub destination: Location,
+	/// The `query_id` field of the `QueryResponse` message.
+	#[codec(compact)]
+	pub query_id: QueryId,
+	/// The `max_weight` field of the `QueryResponse` message.
+	pub max_weight: Weight,
+}
+
+impl TryFrom<OldQueryResponseInfo> for QueryResponseInfo {
+	type Error = ();
+
+	fn try_from(old: OldQueryResponseInfo) -> result::Result<Self, Self::Error> {
+		Ok(Self {
+			destination: old.destination.try_into()?,
+			query_id: old.query_id,
+			max_weight: old.max_weight,
+		})
+	}
+}
+
+/// Contextual data pertaining to a specific list of XCM instructions.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug)]
+pub struct XcmContext {
+	/// The current value of the Origin register of the `XCVM`.
+	pub origin: Option<Location>,
+	/// The identity of the XCM; this may be a hash of its versioned encoding but could also be
+	/// a high-level identity set by an appropriate barrier.
+	pub message_id: XcmHash,
+	/// The current value of the Topic register of the `XCVM`.
+	pub topic: Option<[u8; 32]>,
+}
+
+impl XcmContext {
+	/// Constructor which sets the message ID to the supplied parameter and leaves the origin and
+	/// topic unset.
+	pub fn with_message_id(message_id: XcmHash) -> XcmContext {
+		XcmContext { origin: None, message_id, topic: None }
+	}
+}
+
+/// Cross-Consensus Message: A message from one consensus system to another.
+///
+/// Consensus systems that may send and receive messages include blockchains and smart contracts.
+///
+/// All messages are delivered from a known *origin*, expressed as a `Location`.
+///
+/// This is the inner XCM format and is version-sensitive. Messages are typically passed using the
+/// outer XCM format, known as `VersionedXcm`.
+#[derive(
+	Derivative,
+	Encode,
+	Decode,
+	TypeInfo,
+	xcm_procedural::XcmWeightInfoTrait,
+	xcm_procedural::Builder,
+)]
+#[derivative(Clone(bound = ""), Eq(bound = ""), PartialEq(bound = ""), Debug(bound = ""))]
+#[codec(encode_bound())]
+#[codec(decode_bound())]
+#[scale_info(bounds(), skip_type_params(Call))]
+pub enum Instruction<Call> {
```

Comment:
@joepetrowski @franciscoaguirre 
Joe mentioned [here](https://github.com/polkadot-fellows/runtimes/issues/68#issuecomment-1852027708) `UnpaidTeleport` for XCMv4, how is that? XCMv5 or it is missing here?

Reply:
@franciscoaguirre and I discussed in DM, will make a separate PR for that instruction.

Reply:
I wanted to get this PR merged before thinking of new instructions as there's already a lot to review/audit and conflicts pile on very easily.

There's also a [new Publish instruction](https://github.com/paritytech/polkadot-sdk/pull/2359) that we want, but I'd do it after this PR.